rust/datafusion/src/execution/physical_plan/expressions.rs

```diff
@@ -51,3 +59,163 @@ impl PhysicalExpr for Column {
         Ok(batch.column(self.index).clone())
     }
 }
+
+/// Create a column expression
+pub fn col(i: usize) -> Arc<dyn PhysicalExpr> {
+    Arc::new(Column::new(i))
+}
+
+/// SUM aggregate expression
+pub struct Sum {
+    expr: Arc<dyn PhysicalExpr>,
+}
+
+impl Sum {
+    /// Create a new SUM aggregate function
+    pub fn new(expr: Arc<dyn PhysicalExpr>) -> Self {
+        Sum { expr }
+    }
+}
+
+impl AggregateExpr for Sum {
+    fn name(&self) -> String {
+        "SUM".to_string()
+    }
+
+    fn data_type(&self, input_schema: &Schema) -> Result<DataType> {
+        match self.expr.data_type(input_schema)? {
+            DataType::Int8 | DataType::Int16 | DataType::Int32 | DataType::Int64 => {
+                Ok(DataType::Int64)
+            }
+            DataType::UInt8 | DataType::UInt16 | DataType::UInt32 | DataType::UInt64 => {
+                Ok(DataType::UInt64)
+            }
+            other => Ok(other.clone()),
+        }
+    }
+
+    fn create_accumulator(&self) -> Rc<RefCell<dyn Accumulator>> {
+        Rc::new(RefCell::new(SumAccumulator {
+            expr: self.expr.clone(),
+            sum: None,
+        }))
+    }
+
+    fn create_combiner(&self, column_index: usize) -> Arc<dyn AggregateExpr> {
+        Arc::new(Sum::new(Arc::new(Column::new(column_index))))
+    }
+}
+
+macro_rules! sum_accumulate {
+    ($SELF:ident, $ARRAY:ident, $ROW_INDEX:expr, $ARRAY_TYPE:ident, $SCALAR_VARIANT:ident, $TY:ty) => {{
+        if let Some(array) = $ARRAY.as_any().downcast_ref::<$ARRAY_TYPE>() {
+            let value = array.value($ROW_INDEX);
```

Comment:
You are not handling `NULL` slots here right?  Shouldn't you use [sum](https://github.com/apache/arrow/blob/master/rust/arrow/src/compute/array_ops.rs#L75) from kernels module?  That way `NULL` handling is taken care of.  (SIMD, eventually, will be leveraged automatically as well).

Reply:
This method is called to update an accumulator with the value from a single row within the batch, so SIMD can't help with this particular usage.

Basically I'm building a hash map of accumulators by grouping key, then iterating over the rows in the batch to look up the accumulator based on the grouping key in each row, then updating is.

There is a future optimization that can be made for aggregate queries without grouping keys e.g. `SELECT SUM(a) FROM b` where we could use the sum from kernels to sum entire batches.

Reply:
Ok, I see. Yes, the example without grouping is what I had in mind.  Optimizations can be added later but are you handling `NULL` slots correctly here?

Reply:
Good catch. I was not handling null slots correctly. I have pushed the fix and added a test with all null values.